File: Deploy/RapierLoom.Examples/05-Authentication/CS/AuthenticationAspect.cs
// ----------------------------------------------------------------
// * The LOOM.NET project http://rapier-loom.net *
// (C) Copyright by Wolfgang Schult. All rights reserved. 
// This code is licensed under the Apache License 2.0
// To get more information, go to http://loom.codeplex.com/license
// ----------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Text;

using Loom;
using Loom.JoinPoints;
using System.Net;

namespace Authentication
{
    /// <summary>
    /// This Authentication aspect shows, how to implement an authentication aspect to control the access of 
    /// application methods, simply by annotating the target methods with this aspect. The aspect checks 
    /// the entered username and password and caches the credentials for reuse.
    /// </summary>
    public class AuthenticationAspect : AspectAttribute
    {
        // caches the user credentials
        private NetworkCredential credential;

        [IncludeAll]            // includes all methods
        [Call(Advice.Around)]   // aspect method is called instead of the target method
        public T AuthenticateAction<T>([JPContext] Context context, params object[] parameters)
        {
            Console.WriteLine("*** Action needs authentication.");

            // check for already cached credentials
            if (credential == null)
            {
                Console.Write("Username: ");
                string username = Console.ReadLine();
                Console.Write("Password: ");
                string password = Console.ReadLine();
                credential = new NetworkCredential(username, password);  // store credentials to aspect
            }
            
            if (Authenticator.Authenticate(credential))     // authenticate the user
                return (T)context.Invoke(parameters);       // invoke the original method and return the result
            else
                throw new AuthenticationException("Error: Authentication failed.");  // or throw an authentication exception
        }
        
    }
}

File: Deploy/RapierLoom.Examples/05-Authentication/VB/AuthenticationAspect.vb
Public Class AuthenticationAspect
    Inherits AspectAttribute

    Private credential As Net.NetworkCredential

    <IncludeAll(), [Call](Advice.Around)> _
    Public Function AuthenticateAction(Of T)(<JPContext()> ByVal context As Context, ByVal ParamArray parameters As Object()) As T
        Console.WriteLine("*** Action needs authentication.")
        If (Me.credential Is Nothing) Then
            Console.Write("Username: ")
            Dim username As String = Console.ReadLine
            Console.Write("Password: ")
            Dim password As String = Console.ReadLine
            Me.credential = New Net.NetworkCredential(username, password)
        End If

        If Not Authenticator.Authenticate(Me.credential) Then
            Throw New AuthenticationException("Error: Authentication failed.")
        End If

        Return DirectCast(context.Invoke(parameters), T)
    End Function



End Class


